superset/assets/src/visualizations/wordcloud/buildQuery.ts

```diff
@@ -0,0 +1,15 @@
+import buildQueryContext from 'src/query';
+import { FormData } from './formData';
+
+export default function buildQuery(formData: FormData) {
```

Comment:
I thought about this more as someone who needs to write plugin. Might be cleaner from plugin developer standpoint if developer only has to write

```js
export default function buildQuery(formData: FormData) {
  return buildQueryContext(formData, (baseQueryObject) => [{
     ...baseQueryObject,
     groupby: [formData.series],
  }]);
}
```

Assuming it will always be the same process for  `datasource` and the plugin should only touch the `queries`.

Reply:
Even this would become slightly more complex when modifying the `queries` tho, at that point they seem pretty similar to me?

```javascript
export default function buildQuery(formData: FormData) {
  return buildQueryContext(formData, (baseQueryObject) => {
    const { queries, ...rest } = baseQueryObject;
    const [ query ] = queries;

    return {
      ...rest,
      queries: [{
        ...query,
        groupby: [formData.series],
      }],
    };
  };
}
```

Reply:
Umm, I mean to abstract that complex part into `buildQueryContext`, so vis-specific `buildQuery` 
- does not need to touch `datasource`. 
- does not modify `queries` but receive `baseQueryObject` that is the single `queryObject` currently returned in `queryContext.queries` and use that to construct `queries` and return. 

Current `buildQueryContext`

```js
function buildQueryContext(formData: FormData): QueryContext {
  return {
    datasource: buildDatasource(formData),
    queries: buildQueries(formData),
  };
}
```

With my proposed change

```js
function buildQueryContext(formData: FormData, buildQueries: Function): QueryContext {
  return {
    datasource: buildDatasource(formData),
    queries: buildQueries(buildQueryObject(formData)),
  };
}
```

Reply:
I like these guiding principles and the usage of higher-order function in the proposal:
> vis-specific buildQuery
> - does not need to touch datasource.
> - does not modify queries but receive baseQueryObject that is the single queryObject currently > in queryContext.queries and use that to construct queries and return.

Relating to the `BuildQueryRegistery` discussion, this approach would work nicely (we would register each viz-specific `buildQuery` function).

One small tweak I'm thinking is to define `BaseQueryObject` as a class (instead of a plain js object) so we could more easily add utility functions viz-specific `buildQuery` may need to enrich their queries. For example, imagine generating filter specific structures that can be accessed by viz-specific `buildQuery`s to construct their own pre-processing queries. We would have `.toObject()` in the class as well for accessing the plain js object.

Reply:
Okay, I deciding against the small tweak as I'm writing it. I think we can add it once the need arises. Going to leave baseQueryObject as an interface for now.

Reply:
LGTM